File: app/src/main/java/com/machiav3lli/backup/items/SchedulerItemX.java
/*
 * OAndBackupX: open-source apps backup and restore app.
 * Copyright (C) 2020  Antonios Hazim
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.machiav3lli.backup.items;

import android.view.View;

import androidx.appcompat.widget.AppCompatCheckBox;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.appcompat.widget.LinearLayoutCompat;

import com.machiav3lli.backup.R;
import com.machiav3lli.backup.schedules.HandleAlarms;
import com.machiav3lli.backup.schedules.db.Schedule;
import com.machiav3lli.backup.utils.ItemUtils;
import com.mikepenz.fastadapter.FastAdapter;
import com.mikepenz.fastadapter.items.AbstractItem;

import org.jetbrains.annotations.NotNull;

import java.util.List;

public class SchedulerItemX extends AbstractItem<SchedulerItemX.ViewHolder> {
    Schedule schedule;

    public SchedulerItemX(Schedule schedule) {
        this.schedule = schedule;
    }

    public Schedule getSchedule() {
        return schedule;
    }

    @Override
    public int getLayoutRes() {
        return R.layout.item_scheduler_x;
    }

    @NotNull
    @Override
    public ViewHolder getViewHolder(@NotNull View view) {
        return new ViewHolder(view);
    }

    @Override
    public long getIdentifier() {
        return ItemUtils.calculateScheduleID(schedule);
    }

    @Override
    public int getType() {
        return R.id.fastadapter_item;
    }

    protected static class ViewHolder extends FastAdapter.ViewHolder<SchedulerItemX> {
        AppCompatCheckBox checkbox = itemView.findViewById(R.id.enableCheckbox);
        AppCompatTextView timeLeft = itemView.findViewById(R.id.timeLeft);
        LinearLayoutCompat timeLeftLine = itemView.findViewById(R.id.timeLeftLine);
        AppCompatTextView scheduleMode = itemView.findViewById(R.id.schedMode);
        AppCompatTextView scheduleSubMode = itemView.findViewById(R.id.schedSubMode);

        public ViewHolder(View view) {
            super(view);
        }

        @Override
        public void bindView(@NotNull SchedulerItemX item, @NotNull List<?> list) {
            final Schedule schedule = item.getSchedule();
            setScheduleMode(schedule);
            setScheduleSubMode(schedule);
            checkbox.setChecked(schedule.getEnabled());
            setTimeLeft(schedule, System.currentTimeMillis());
            final long tag = schedule.getId();
            checkbox.setTag(tag);
        }

        void setScheduleMode(Schedule schedule) {
            switch (schedule.getMode().getValue()) {
                case 1:
                    scheduleMode.setText(R.string.radio_user);
                    break;
                case 2:
                    scheduleMode.setText(R.string.radio_system);
                    break;
                case 3:
                    scheduleMode.setText(R.string.showNewAndUpdated);
                    break;
                default:
                    scheduleMode.setText(R.string.radio_all);
                    break;
            }
        }

        void setScheduleSubMode(Schedule schedule) {
            switch (schedule.getSubMode().getValue()) {
                case 1:
                    scheduleSubMode.setText(R.string.radio_apk);
                    break;
                case 2:
                    scheduleSubMode.setText(R.string.radio_data);
                    break;
                default:
                    scheduleSubMode.setText(R.string.radio_both);
                    break;
            }
        }

        void setTimeLeft(Schedule schedule, long now) {
            if (!schedule.getEnabled()) {
                timeLeft.setText("");
                timeLeftLine.setVisibility(View.INVISIBLE);
            } else {
                final long timeDiff = HandleAlarms.timeUntilNextEvent(schedule.getInterval(),
                        schedule.getTimeHour(), schedule.getTimeMinute(), schedule.getTimePlaced(), now);
                int sum = (int) (timeDiff / 1000f / 60f);
                int hours = sum / 60;
                int minutes = sum % 60;
                timeLeft.setText(String.format(" %s:%s", hours, minutes));
                timeLeftLine.setVisibility(View.VISIBLE);
            }
        }

        @Override
        public void unbindView(@NotNull SchedulerItemX item) {
            scheduleMode.setText(null);
            scheduleSubMode.setText(null);
            timeLeft.setText(null);
        }
    }
}

File: app/src/main/java/com/machiav3lli/backup/items/SchedulerItemX.kt
/*
 * OAndBackupX: open-source apps backup and restore app.
 * Copyright (C) 2020  Antonios Hazim
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.machiav3lli.backup.items

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.machiav3lli.backup.*
import com.machiav3lli.backup.databinding.ItemSchedulerXBinding
import com.machiav3lli.backup.dbs.Schedule
import com.machiav3lli.backup.utils.modeToModes
import com.machiav3lli.backup.utils.setExists
import com.machiav3lli.backup.utils.timeUntilNextEvent
import com.mikepenz.fastadapter.binding.AbstractBindingItem
import com.mikepenz.fastadapter.diff.DiffCallback
import java.time.LocalTime
import java.util.concurrent.TimeUnit
import kotlin.math.abs

class SchedulerItemX(var schedule: Schedule) : AbstractBindingItem<ItemSchedulerXBinding>() {

    override var identifier: Long
        get() = schedule.id
        set(identifier) {
            super.identifier = identifier
        }

    override val type: Int
        get() = R.id.fastadapter_item

    override fun createBinding(
        inflater: LayoutInflater,
        parent: ViewGroup?
    ): ItemSchedulerXBinding {
        return ItemSchedulerXBinding.inflate(inflater, parent, false)
    }

    override fun bindView(binding: ItemSchedulerXBinding, payloads: List<Any>) {
        binding.schedName.text = schedule.name
        binding.systemFilter.setExists(schedule.filter and MAIN_FILTER_SYSTEM == MAIN_FILTER_SYSTEM)
        binding.userFilter.setExists(schedule.filter and MAIN_FILTER_USER == MAIN_FILTER_USER)
        binding.updatedFilter.setExists(schedule.specialFilter == SPECIAL_FILTER_NEW_UPDATED)
        binding.launchableFilter.setExists(schedule.specialFilter == SPECIAL_FILTER_LAUNCHABLE)
        binding.oldFilter.setExists(schedule.specialFilter == SPECIAL_FILTER_OLD)
        modeToModes(schedule.mode).let {
            binding.apkMode.setExists(it.contains(MODE_APK))
            binding.dataMode.setExists(it.contains(MODE_DATA))
            binding.deDataMode.setExists(it.contains(MODE_DATA_DE))
            binding.extDataMode.setExists(it.contains(MODE_DATA_EXT))
            binding.obbMode.setExists(it.contains(MODE_DATA_OBB))
        }
        binding.enableCheckbox.isChecked = schedule.enabled
        setTimeLeft(binding)
    }

    override fun unbindView(binding: ItemSchedulerXBinding) {
        binding.timeLeft.text = null
        binding.schedName.text = null
    }

    private fun setTimeLeft(binding: ItemSchedulerXBinding) {
        if (!schedule.enabled) {
            binding.timeLeft.text = ""
            binding.timeLeftLine.visibility = View.INVISIBLE
        } else {
            val timeDiff = abs(timeUntilNextEvent(schedule, System.currentTimeMillis()))
            val days = TimeUnit.MILLISECONDS.toDays(timeDiff).toInt()
            if (days == 0) {
                binding.daysLeft.visibility = View.GONE
            } else {
                binding.daysLeft.visibility = View.VISIBLE
                binding.daysLeft.text = binding.root.context.resources.getQuantityString(
                    R.plurals.days_left,
                    days,
                    days
                )
            }
            val hours = TimeUnit.MILLISECONDS.toHours(timeDiff).toInt() % 24
            val minutes = TimeUnit.MILLISECONDS.toMinutes(timeDiff).toInt() % 60
            binding.timeLeft.text = LocalTime.of(hours, minutes).toString()
            binding.timeLeftLine.visibility = View.VISIBLE
        }
    }

    companion object {
        class SchedulerDiffCallback : DiffCallback<SchedulerItemX> {
            override fun areContentsTheSame(
                oldItem: SchedulerItemX,
                newItem: SchedulerItemX
            ): Boolean {
                return oldItem.schedule.id == newItem.schedule.id
            }

            override fun areItemsTheSame(
                oldItem: SchedulerItemX,
                newItem: SchedulerItemX
            ): Boolean {
                return oldItem.schedule.enabled == newItem.schedule.enabled
                        && oldItem.schedule.name == newItem.schedule.name
                        && oldItem.schedule.timeHour == newItem.schedule.timeHour
                        && oldItem.schedule.timeMinute == newItem.schedule.timeMinute
                        && oldItem.schedule.interval == newItem.schedule.interval
                        && oldItem.schedule.timePlaced == newItem.schedule.timePlaced
                        && oldItem.schedule.filter == newItem.schedule.filter
                        && oldItem.schedule.mode == newItem.schedule.mode
                        && oldItem.schedule.specialFilter == newItem.schedule.specialFilter
                        && oldItem.schedule.customList == newItem.schedule.customList
                        && oldItem.schedule.blockList == newItem.schedule.blockList
            }

            override fun getChangePayload(
                oldItem: SchedulerItemX,
                oldItemPosition: Int,
                newItem: SchedulerItemX,
                newItemPosition: Int
            ): Any? {
                return null
            }
        }
    }
}